interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const posixMQSummary = `allows access to POSIX message queues`
+
+// This interface is super-privileged
+const posixMQBaseDeclarationSlots = `
+  posix-mq:
+    allow-installation: false
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+// Paths can only be specified by the slot and a slot needs to exist for the
+// plug to connect to, so the plug does not need to be marked super-privileged
+const posixMQBaseDeclarationPlugs = `
+  posix-mq:
+    allow-installation: true
+    allow-connection:
+      slot-attributes:
+        posix-mq: $PLUG(posix-mq)
+    allow-auto-connection:
+      slot-publisher-id:
+        - $PLUG_PUBLISHER_ID
+      slot-attributes:
+        posix-mq: $PLUG(posix-mq)
+`
+
+const posixMQPermanentSlotSecComp = `
+mq_open
+mq_getsetattr
+mq_unlink
+mq_notify
+mq_timedreceive
+mq_timedsend
+`
+
+var posixMQPlugPermissions = []string{
+	"open",
+	"read",
+	"write",
+	"create",
+	"delete",
+}
+
+var posixMQDefaultPlugPermissions = []string{
+	"read",
+	"write",
+}
+
+// Ensure that the name matches the criteria from the mq_overview man page:
+//   Each message queue is identified by a name of the form /somename;
+//   that is, a null-terminated string of up to NAME_MAX (i.e., 255)
+//   characters consisting of an initial slash, followed by one or more
+//   characters, none of which are slashes.
+var posixMQNamePattern = regexp.MustCompile(`^/[^/]{1,255}$`)
+
+type posixMQInterface struct {
+	commonInterface
+}
+
+func (iface *posixMQInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              posixMQSummary,
+		BaseDeclarationSlots: posixMQBaseDeclarationSlots,
+		BaseDeclarationPlugs: posixMQBaseDeclarationPlugs,
+	}
+}
+
+func (iface *posixMQInterface) Name() string {
+	return "posix-mq"
+}
+
+func (iface *posixMQInterface) checkPosixMQAppArmorSupport() error {
+	if apparmor_sandbox.ProbedLevel() == apparmor_sandbox.Unsupported {
+		/* AppArmor is not supported at all; no need to add rules */
+		return nil
+	}
+
+	features, err := apparmor_sandbox.ParserFeatures()
+	if err != nil {
+		return err
+	}
+
+	if !strutil.ListContains(features, "mqueue") {
+		return fmt.Errorf("AppArmor does not support POSIX message queues - cannot setup or connect interfaces")
+	}
+
+	return nil
+}
+
+func (iface *posixMQInterface) isValidPermission(perm string) bool {
+	for _, validPerm := range posixMQPlugPermissions {
+		if perm == validPerm {
+			return true
+		}
+	}
+	return false
+}
+
+func (iface *posixMQInterface) validatePermissionList(perms []string, name string) error {
+	for _, perm := range perms {
+		if !iface.isValidPermission(perm) {
+			return fmt.Errorf("posix-mq slot %s permission \"%s\" not valid, must be one of %v", name, perm, posixMQPlugPermissions)
```

Comment:
IIRC, in another review @pedronis suggested not to print the slot name in the error messages.

Reply:
Should I remove the names?

Reply:
yes, higher levels will add the slot name to the error.  Also this and many other error paths are not tested.